### ISSUE TITLE
Question.

### ISSUE THREAD

**Issue by cerelli** [spatie/laravel-event-projector#69]
Not knowing it, I wanted to know if this package can replace or do the same things as this: https://github.com/owen-it/laravel-auditing
Thank you.

Marco

**Comment by 36864** [comment]
They fulfill different requirements.
Auditing takes a traditional MVC application and adds logging to all events that happen on the models. This is something that you can add to an application at any time and start gathering audit logs.

This package helps you implement event sourcing, which means that instead of handling models, your app will deal in events, which can then be processed into models. This is fundamentally different from the regular MVC pattern that you may be used to working with in laravel. You can't just "work it into" and existing app.


With auditing, every time you update a Post, you get a log with metadata about that update. This could look something like `Post 42 was updated by User 1.  The Topic was changed from "Foo" to "Bar!"`.
If you want to know what the topic was before this change, you can check the latest entry in your audit log and it's right there.

With event sourcing, when you want to update a Post, you emit an event that describes that update, and then a projector will update the Post model accordingly. This could look something like "Post 42 was updated by User 1. The Topic is now "Bar!"`. If you want to know what the topic was before this change, you have to find the last event that changed the topic (or created the post) before this one.


With auditing, your model is changed, and then an entry in the audit log is created to describe that change.

With event sourcing, an event is logged that describes a change, and then the model is updated to reflect that change.


The event source log can be used as an audit log by definition. Everything that happens in your application is an event, and the state of your application is a product of all those events being processed sequentially. However, this isn't something you can just drop into an existing project to replace existing audit mechanisms.

I've been describing a very basic use case here. It's worth noting that you can have many projectors handling the same events, so there isn't a 1-to-1 mapping between an event happening and a model being changed. 
For example, you could have a projector that tracks statistics for your app, incrementing a counter every time a "Post Created" event is processed. This would trigger changes to two models: posts and statistics. Conversely, with auditing, you would have a statistics model that keeps a log of when and why it was updated, which is separate from the audit logs of the posts that trigger those updates.

I hope I was clear in explaining some of the differences here.
If you want to see a basic implementation of this, I built a simple task list app to teach myself event sourcing: https://github.com/36864/Event-Sourced-Task-Lists

**Comment by tucq88** [comment]
Why you're so awesome @36864 ????? Great comment, great repo!
I'm on very first steps to learn and implement Event Sourced apps. Could you suggest me on what to read and how to start implement Event Sourced apps ? 

Thanks so so so so so much

**Comment by 36864** [comment]
@tucq88 To build my little sample app, I mostly used [this repo's documentation](https://docs.spatie.be/laravel-event-projector/v1/basic-usage/writing-your-first-projector) (I just substituted accounts and money for task lists and tasks) and the [article that's linked in the readme](https://kickstarter.engineering/event-sourcing-made-simple-4a2625113224). Note that there are some differences in terminology (Projector <=> Calculator, Aggregate <=> Model) and in implementation (separate event tables per model vs single events table).

As for implementing your first event sourced app, I would recommend trying to replicate a simple app that you've made in the past, so you know what _should_ be happening in the app. Take it slow, start with the basics of just creating and deleting a model, and then think of what actions you want to be able to perform on those models. For every action you identify, implement it one at a time by creating a new event and making sure it works before implementing the next action.

When I started my first app it went something like this:
- Make an event to create a task
- Create a task
- Make an event to delete a task
- Delete the task I created earlier
- Make an event to mark a task as complete
and so on and so on. (not pictured: about two full days of fumbling about with concepts I didn't fully understand)

I am by no means an expert on the subject, and there's always a chance I'm completely wrong about everything I've said.

The best advice I have for you is to just give it a shot. It might take you a few tries before you get it right, or it might just "click" and make sense for you right off the bat. No way to know until you try.
